File: src/main/java/com/github/zyypj/tadeuBooter/api/downloaders/dependencies/Dependency.java
package com.github.zyypj.tadeuBooter.api.downloaders.dependencies;

import com.google.common.base.Stopwatch;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.java.JavaPlugin;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.concurrent.CompletableFuture;

/**
 * Essa classe foi inspirada numa implementação já existente!
 *
 * @author syncwrld (github.com/syncwrld)
 */
@Getter
@RequiredArgsConstructor
public class Dependency {
    @Setter
    private static JavaPlugin plugin;

    @Setter
    private static String prefix = "§8§l[DependencyLoader] §f";

    private final DependencyInfo dependencyInfo;
    private long bytesSize = 0L;

    public CompletableFuture<Void> download() {
        return CompletableFuture.runAsync(() -> {
            if (plugin == null) {
                throw new IllegalStateException("O plugin não foi definido. Use Dependency.setPlugin(JavaPlugin) antes de usar essa função.");
            }

            if (Bukkit.getPluginManager().getPlugin(dependencyInfo.getName()) != null) return;

            File pluginFile = new File(plugin.getDataFolder().getParentFile(),
                    dependencyInfo.getName().toLowerCase() + "-" + dependencyInfo.getVersion() + ".jar");

            if (pluginFile.exists()) {
                log("§eDependência já existe: " + dependencyInfo.getName() + " v" + dependencyInfo.getVersion(), true);
                return;
            }

            Stopwatch stopwatch = Stopwatch.createStarted();
            try {
                HttpURLConnection connection = (HttpURLConnection) new URL(dependencyInfo.getDownloadURL()).openConnection();
                connection.setRequestProperty("User-Agent", "Mozilla/5.0");
                connection.setRequestMethod("GET");
                connection.setConnectTimeout(10000);
                connection.setReadTimeout(30000);
                connection.connect();

                try (InputStream inputStream = connection.getInputStream()) {
                    Files.copy(inputStream, pluginFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
                }

                this.bytesSize = pluginFile.length();
                log(String.format("§aDownloaded: %s v%s - Time: %s - Size: %s",
                        dependencyInfo.getName(), dependencyInfo.getVersion(), stopwatch, asReadableSize(this.bytesSize)), true);

                plugin.getServer().getScheduler().runTask(plugin, () -> loadPlugin(pluginFile));

            } catch (IOException e) {
                log(String.format("§cFailed to download %s v%s - Cause: %s",
                        dependencyInfo.getName(), dependencyInfo.getVersion(), e.getMessage()), true);
            }
        });
    }

    private void loadPlugin(File pluginFile) {
        try {
            Plugin loadedPlugin = Bukkit.getPluginManager().loadPlugin(pluginFile);
            if (loadedPlugin != null) {
                Bukkit.getPluginManager().enablePlugin(loadedPlugin);
            }
            if (dependencyInfo.getName().equalsIgnoreCase("PlaceholderAPI")) {
                log("§cPlaceholderAPI installed! Restart the server to avoid errors.", true);
            }
        } catch (Exception e) {
            log(String.format("§cFailed to start dependency: %s - Cause: %s", dependencyInfo.getName(), e.getMessage()), true);
        }
    }

    private void log(String message, boolean debug) {
        if (plugin == null) {
            throw new IllegalStateException("O plugin não foi definido. Use Dependency.setPlugin(JavaPlugin) antes de usar essa função.");
        }

        message = ChatColor.translateAlternateColorCodes('&', message);

        if (debug && plugin.getConfig().getBoolean("debug", true)) {
            Bukkit.getConsoleSender().sendMessage(prefix + message);
            return;
        }

        Bukkit.getConsoleSender().sendMessage(message);
    }

    private static String asReadableSize(long byteSize) {
        if (byteSize < 1024) return byteSize + " B";
        if (byteSize < 1048576) return byteSize / 1024 + " KB";
        if (byteSize < 1073741824) return byteSize / 1048576 + " MB";
        return byteSize / 1073741824 + " GB";
    }
}

File: spigot/src/main/java/me/zyypj/booter/minecraft/spigot/dependency/Dependency.kt
package me.zyypj.booter.minecraft.spigot.dependency

import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.util.*
import java.util.concurrent.CompletableFuture

/**
 * Representa uma dependência a baixar e carregar dinamicamente.
 */
class Dependency(val dependencyInfo: DependencyInfo) {

    companion object {
        private var plugin: JavaPlugin? = null
        private var prefix: String = "§8§l[DependencyLoader] §f"

        /** Define o plugin que gerencia os downloads (deve ser chamado antes). */
        @JvmStatic
        fun setPlugin(p: JavaPlugin) {
            plugin = p
        }

        /** Prefixo para logs. */
        @JvmStatic
        fun setPrefix(p: String) {
            prefix = p
        }
    }

    var bytesSize: Long = 0L
        private set

    /** Inicia download assíncrono. */
    fun download(): CompletableFuture<Void> = CompletableFuture.runAsync {
        val pl = plugin ?: throw IllegalStateException(
            "O plugin não foi definido. Use Dependency.setPlugin() antes."
        )
        if (Bukkit.getPluginManager().getPlugin(dependencyInfo.name) != null) return@runAsync

        val outFile = File(
            pl.dataFolder.parentFile,
            "${dependencyInfo.name.lowercase(Locale.getDefault())}-${dependencyInfo.version}.jar"
        )
        if (outFile.exists()) {
            log(pl, "§eDependência já instalada: ${dependencyInfo.name} v${dependencyInfo.version}", true)
            return@runAsync
        }

        val stopwatch = System.nanoTime()
        try {
            val conn = URL(dependencyInfo.downloadURL).openConnection() as HttpURLConnection
            conn.setRequestProperty("User-Agent", "Mozilla/5.0")
            conn.requestMethod = "GET"
            conn.connectTimeout = 10000
            conn.readTimeout = 30000
            conn.connect()
            conn.inputStream.use { ins ->
                Files.copy(ins, outFile.toPath(), StandardCopyOption.REPLACE_EXISTING)
            }
            bytesSize = outFile.length()
            val elapsed = (System.nanoTime() - stopwatch) / 1_000_000
            log(
                pl, "§aBaixado: ${dependencyInfo.name} v${dependencyInfo.version} - Tempo: ${elapsed}ms - Tamanho: ${
                    asReadableSize(
                        bytesSize
                    )
                }", true
            )
            pl.server.scheduler.runTask(pl) { loadPlugin(outFile) }
        } catch (e: IOException) {
            log(
                pl, "§cFalha ao baixar ${dependencyInfo.name} v${dependencyInfo.version} - Causa: ${e.message}", true
            )
        }
    }

    private fun loadPlugin(file: File) {
        try {
            val loaded = Bukkit.getPluginManager().loadPlugin(file)
            if (loaded != null) Bukkit.getPluginManager().enablePlugin(loaded)
            if (dependencyInfo.name.equals("PlaceholderAPI", ignoreCase = true)) {
                log(
                    plugin!!, "§cPlaceholderAPI instalado! Reinicie o servidor para evitar erros.", true
                )
                plugin!!.server.dispatchCommand(
                    plugin!!.server.consoleSender, "restart"
                )
            }
        } catch (e: Exception) {
            log(
                plugin!!, "§cFalha ao iniciar dependência: ${dependencyInfo.name} - Causa: ${e.message}", true
            )
        }
    }

    private fun log(pl: JavaPlugin, message: String, debug: Boolean) {
        val msg = ChatColor.translateAlternateColorCodes('&', message)
        if (debug && pl.config.getBoolean("debug", true)) {
            Bukkit.getConsoleSender().sendMessage(prefix + msg)
        } else {
            Bukkit.getConsoleSender().sendMessage(msg)
        }
    }

    private fun asReadableSize(bytes: Long): String = when {
        bytes < 1024 -> "$bytes B"
        bytes < 1_048_576 -> "${bytes / 1024} KB"
        bytes < 1_073_741_824 -> "${bytes / 1_048_576} MB"
        else -> "${bytes / 1_073_741_824} GB"
    }
}